File: scala/src/main/scala/day2/SolutionA.scala
package day2

import java.nio.file.{Files, Paths}

object SolutionA extends App {

  sealed trait Choice

  object Choice {
    case object Rock extends Choice

    case object Paper extends Choice

    case object Scissors extends Choice
  }

  val columnAMapping: Map[String, Choice] = Map(
    "A" -> Choice.Rock,
    "B" -> Choice.Paper,
    "C" -> Choice.Scissors
  )

  val columnBMapping: Map[String, Choice] = Map(
    "X" -> Choice.Rock,
    "Y" -> Choice.Paper,
    "Z" -> Choice.Scissors
  )

  val shapeScores: Map[Choice, Int] = Map(
    Choice.Rock     -> 1,
    Choice.Paper    -> 2,
    Choice.Scissors -> 3
  )

  def roundScore(me: Choice, them: Choice) = (me, them) match {
    case (Choice.Rock, Choice.Scissors)     => 6
    case (Choice.Rock, Choice.Rock)         => 3
    case (Choice.Rock, Choice.Paper)        => 0
    case (Choice.Paper, Choice.Rock)        => 6
    case (Choice.Paper, Choice.Paper)       => 3
    case (Choice.Paper, Choice.Scissors)    => 0
    case (Choice.Scissors, Choice.Paper)    => 6
    case (Choice.Scissors, Choice.Scissors) => 3
    case (Choice.Scissors, Choice.Rock)     => 0
  }

  val input = Files.readString(Paths.get("src/main/resources/day2/input_a.txt"))

  val rounds = input
    .split("\n")
    .map(_.split(' '))
    .map { case Array(a, b) => columnAMapping(a) -> columnBMapping(b) }

  val result = rounds
    .map { case (them, me) =>
      val shapeScore = shapeScores(me)
      val score = roundScore(me, them)
      shapeScore + score
    }
    .sum

  println(result) // 11475

}

File: python/aoc_2022/day2/solution_a.py
from enum import Enum

class Choice(Enum):
    ROCK = 1
    PAPER = 2
    SCISSORS = 3

column_a_mappings = {
    "A": Choice.ROCK,
    "B": Choice.PAPER,
    "C": Choice.SCISSORS,
}

column_b_mappings = {
    "X": Choice.ROCK,
    "Y": Choice.PAPER,
    "Z": Choice.SCISSORS,
}

scores = {
    (Choice.ROCK, Choice.SCISSORS): 6,
    (Choice.ROCK, Choice.ROCK): 3,
    (Choice.ROCK, Choice.PAPER): 0,
    (Choice.PAPER, Choice.ROCK): 6,
    (Choice.PAPER, Choice.PAPER): 3,
    (Choice.PAPER, Choice.SCISSORS): 0,
    (Choice.SCISSORS, Choice.PAPER): 6,
    (Choice.SCISSORS, Choice.SCISSORS): 3,
    (Choice.SCISSORS, Choice.ROCK): 0
}
        
def process(line):
    tokens = line.split(" ")
    
    them = column_a_mappings[tokens[0]]
    me = column_b_mappings[tokens[1]]
    
    return me.value + scores[(me, them)]


if __name__ == "__main__":
    
    with open("inputs/day2/input_a.txt", "r") as f:
        input = f.read()
        
    result = sum(map(process, input.split("\n")))
        
    print(result) # 11475